extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultClientTest.java

```diff
@@ -9,228 +9,619 @@
+                            .body(ResponseBody.create(body, MediaType.get("application/json")))
+                            .protocol(Protocol.HTTP_1_1)
+                            .request(new Request.Builder().url("http://any").build())
+                            .build();
+                    when(httpClient.execute(any(Request.class), anyList())).thenReturn(response);
+                    mockedFallbackFactories.when(() -> FallbackFactories.retryWhenStatusIsNotIn(NON_RETRYABLE_STATUS_CODES)).thenCallRealMethod();
+
+                    var tokenLookUpResult = vaultClient.lookUpToken();
+
+                    mockedFallbackFactories.verify(() -> FallbackFactories.retryWhenStatusIsNotIn(NON_RETRYABLE_STATUS_CODES));
+                    var tokenLookUpResponse = tokenLookUpResult.getContent();
+                    assertThat(tokenLookUpResponse.getRequestId()).isEqualTo("585caa70-402d-e5c4-0977-5463105ba9be");
+                    assertThat(tokenLookUpResponse.getLeaseId()).isEqualTo("2346aa70-402d-e114-0977-546fad339fbe");
+                    assertThat(tokenLookUpResponse.getLeaseDuration()).isEqualTo(10000);
+                    assertThat(tokenLookUpResponse.isRenewable()).isFalse();
+                    assertThat(tokenLookUpResponse.getWarnings()).isEmpty();
+                    var data = tokenLookUpResponse.getData();
+                    assertThat(data.getAccessor()).isEqualTo("wbXdLDXxtwKZf7dLErtAq5vz");
+                    assertThat(data.getCreationTime()).isEqualTo(1704989507L);
+                    assertThat(data.getCreationTtl()).isEqualTo(2592000L);
+                    assertThat(data.getDisplayName()).isEqualTo("token");
+                    assertThat(data.getEntityId()).isEqualTo("entityId");
+                    assertThat(data.getExpireTime()).isEqualTo("2024-02-10T17:11:47.881108+01:00");
+                    assertThat(data.getId()).isEqualTo("hvs.CAESIDb3_aAS0fYUVzoLKzHDX3-GWOE6Gy5jMAvWRsfjJhVKGh4KHGh2cy5Oc0lNQWdjUVNjdnhNeWJhUk1zbDM2Wks");
+                    assertThat(data.getIssueTime()).isEqualTo("2024-01-11T17:11:47.881135+01:00");
+                    assertThat(data.getMeta()).isEmpty();
+                    assertThat(data.getNumUses()).isEqualTo(5);
+                    assertThat(data.isOrphan()).isFalse();
+                    assertThat(data.getExplicitMaxTtl()).isEqualTo(2592000L);
+                    assertThat(data.getPath()).isEqualTo("auth/token/create");
+                    assertThat(data.getPeriod()).isEqualTo("1h");
+                    assertThat(data.getPolicies()).isEqualTo(List.of("default", "root"));
+                    assertThat(data.isRenewable()).isTrue();
+                    assertThat(data.getTtl()).isEqualTo(2589147L);
+                    assertThat(data.getType()).isEqualTo("service");
+                }
+            }
+
+            @ParameterizedTest
+            @ValueSource(ints = {400, 403, 404, 405, 412, 429, 472, 473, 500, 501, 502, 503})
```

Comment:
I think we're putting too much detail in here, with the `FallbackFactory` abstraction we have the possibility to split the concerns. Personally I would create a `HashicorpVaultClientFallbackFactory` that implements the retry logic when specific status are returned, it can be tested separately (please look at `EdcHttpClientImplTest.execute_fallback_shouldRetryIfStatusIsNot2xxOr4xx` as an example).

Then here that FallbackFactory will be injected and mocked: we can then simply verify its usage.